src/codegen.cpp

```diff
@@ -2510,7 +2519,10 @@ static bool emit_builtin_call(jl_cgval_t *ret, jl_value_t *f, jl_value_t **args,
             if (jl_is_leaf_type(tp0)) {
                 emit_expr(args[1], ctx);
                 assert(jl_is_datatype(tp0));
-                *ret = mark_julia_type(ConstantInt::get(T_size, jl_datatype_nfields(tp0)), false, jl_long_type, ctx);
+                jl_value_t *c = jl_box_long(jl_datatype_nfields(tp0));
+                if (ctx->linfo->def) // toplevel exprs are already rooted
+                    jl_add_linfo_root(ctx->linfo, c);
```

Comment:
So `mark_julia_const` is essentially (partially) unusable in this case? With inference inable to infer `sizeof`, I really need codegen to know about this intrinsic being constant.
What approach would you merge? Rooting into a `jl_codectx_t` array, as @carnaval suggested? Or only marking const in the case of non-toplevel code?